### ISSUE TITLE
dotnet Template

### ISSUE THREAD

**Issue by kMutagene** [CSBiology/CsbScaffold#3]
Creating a dotnet template for this project would be a big step forward, as this fixes both distribution and keeping up-to-date with the current scaffold. Some thoughts on the template:

### The Name

Maybe it is not good to keep CSB in the name. Something like `compBio` or `bioProject` maybe? I dont think that 'CSBScaffold` makes it obvious for anyone but us what the project is about.

My current favourite would be `dotnet new bioProject`

### What should the template install?

Basically two ways to go here, one can get rid of the initial scaffold structure, meaning for every new project you would use `dotnet new`. Or we create the whole scaffold with some user settings using `dotnet new` and then use FAKE to create new project folders (this is the way i implemenmted in in the current v2 branch)

### How far should setup go?

We could even go so far and install DOcker via chocolatey if it is not present, but we dont want to create a package manager here, it is just a script project scaffold. Anyways, one has to talk about where to draw the line.

**Comment by kMutagene** [comment]
Paging @muehlhaus , @ZimmerD , @bvenn , @HLWeil , @Joott , @freymaurer for feedback when you have time 😄

**Comment by ZimmerD** [comment]
Regarding the Name: I agree. How about FsBioLab? This would be in line with this popular FSharp lib [https://fslab.org/](url).

**Comment by ZimmerD** [comment]
Regarding the question how far should setup go: Is it possible to create multiple templates?

**Comment by kMutagene** [comment]
@ZimmerD 

> Regarding the question how far should setup go: Is it possible to create multiple templates?

You mean for the same project? im not aware if that's possible, but you can pass arguments after `dotnet new`, so things like `dotnet new FsBioLab --setup full` or something like that would work. what we could do for example is at least check and notify the user if all prerequisites are installed on the system.

**Comment by HLWeil** [comment]
Regarding the name: I agree `FsBioLab` is a good option as it might give people an idea about what this project actually is. Also I think `FsBioSuite` or just `BioSuite`  sounds good.

**Comment by HLWeil** [comment]
> Basically two ways to go here, one can get rid of the initial scaffold structure, meaning for every new project you would use dotnet new. 

So for every research project you would download all dlls again or did I misunterstand you?

> Regarding the question how far should setup go: Is it possible to create multiple templates?


I agree there should be at least some decision making possible. How about at least two templates like `Core` and `Full` ?

@kMutagene you think it will be possible to simply add target templates afterwards with your current planning?

**Comment by kMutagene** [comment]
> So for every research project you would download all dlls again or did I misunterstand you?

Nope. Take a look at the v2 branch. The [createProject target](https://github.com/CSBiology/CsbScaffold/blob/v2/build.fsx#L123) creates the folder structure and references load scripts generated by paket. These are now stored in the global nuget cache so you always have to download them only once at setup and then only if you update the packages. 

The question goes more in the direction of what the actual `dotnet new` template should do, just download the scaffold as folder and you have to run `fake.cmd -t createProject` or if we can manage the command to create a new project folder structure INSIDE the scaffold folder when it is present and download the scaffold first if not.

> Regarding the question how far should setup go: Is it possible to create multiple templates?
> you think it will be possible to simply add target templates afterwards with your current planning?

Just clarifying here, these templates do NOT reflect the package groups. Here i want to discuss if we take care of other external prerequisites (e.g. installing docker for the user) or not. Templates for package groups can be already easiliy added in v2 by [adding a group in the paket.dependencies file](https://github.com/CSBiology/CsbScaffold/blob/v2/paket.dependencies#L29) and [modelling the group in the fake script](https://github.com/CSBiology/CsbScaffold/blob/v2/build.fsx#L42). 

If we want to add multiple installation targets (Like `Full` or `Core`) we have to model that in the dotnet template config file

Edit: and yes you are right this can be added after the initial template is set up but we still have to discuss it
